### PR TITLE
v4.0.0

### PR DESCRIPTION
**Pull request by rezigned**
Prepare for v4.0 release.

- [x] Regenerate codes
- [x] Fix tests
- [x] Schedules
  - [x] Change `ChargeSchedule` and `TransferSchedule` 's implementation from `ISerializable` to `Model` (the former is incorrect).
  - [x] Rename `*Scheduling` to `*Schedule` e.g. `ChargeScheduling` to `ChargeSchedule`
- [x] Consistencies
  - [x] Fix inconsistencies in `v3.0` such as
    -  `Card.UpdateRequestBuilder(cardId, customerId)` contains wrong order of params (it should be `Card.UpdateRequestBuilder(customerId, cardId)`.
    - `Document.GetRequestBuilder(documentId, disputeId)` to `Document.GetRequestBuilder(disputeId, documentId)`
  - [x] Date & Time
    - [x] `Dispute.closedDate` -> `closedAt` (`closed_at` in API)
    - [x] `Receipt.issuedDate` -> `issuedOn` (`issued_on` in API)
    - [x] `Occurrence.scheduleDate` -> `scheduleOn` (`scheduled_on` in API)
    - [x] `Occurrence.retryDate` -> `retryOn` (`retry_on` in API)
    - [x] `Schedule.startDate` -> `startOn` (`start_on` in API)
    - [x] `Schedule.endDate` -> `endOn` (`end_on` in API)
    - [x] `Schedule.nextOccurrencesDates` -> `nextOccurrencesOn` (`next_occurrences_on` in API)

### CODE REVIEW

File: src/main/java/co/omise/models/Link.java [hunk 1]
```diff
@@ -152,50 +231,47 @@ public CreateRequestBuilder multiple(boolean multiple) {
...
+        protected RequestBody payload() throws IOException  {
+            return serialize();
         }
     }
 
-    /**
-     * The {@link RequestBuilder} class for retrieving all Links that belong to an account.
-     */
-    public static class ListRequestBuilder extends RequestBuilder<ScopedList<Link>> {
+    public static class ListChargesRequestBuilder extends RequestBuilder<ScopedList<Charge>> {
```
Comment (nikolack) [inline src/main/java/co/omise/models/Link.java:246]: There are so many different RequestBuilders in this class. Maybe it wouldn't be a bad idea to put a description on what each of them represent.

File: src/main/java/co/omise/models/Refund.java [hunk 2]
```diff
@@ -13,116 +13,198 @@
...
     public static class CreateRequestBuilder extends RequestBuilder<Refund> {
         private String chargeId;
+
         @JsonProperty
         private long amount;
         @JsonProperty
         private Map<String, Object> metadata;
-
+        @JsonProperty("void")
+        private boolean isVoid;
```
Comment (nuxzero) [inline src/main/java/co/omise/models/Refund.java:195]: Looks not legit but can't use `void`. 😂

File: src/main/java/co/omise/models/ChargeStatus.java [hunk 3]
```diff
@@ -3,9 +3,16 @@
...
+    Expired,
+    @JsonProperty("failed")
+    Failed,
+    @JsonProperty("pending")
+    Pending,
+    @JsonProperty("reversed")
+    Reversed,
+    @JsonProperty("successful")
+    Successful,
+    @JsonProperty("unknown")
```
[the remark this answers is not in the document]
Comment (rezigned) [reply src/main/java/co/omise/models/ChargeStatus.java:16]: yes, it's new status

File: src/main/java/co/omise/models/Customer.java [hunk 4]
```diff
@@ -13,136 +14,100 @@
...
+ * @see <a href="https://www.omise.co/customers-api">Customer API</a>
  */
 public class Customer extends Model {
+    private ScopedList<Card> cards;
     @JsonProperty("default_card")
     private String defaultCard;
-    private String email;
     private String description;
+    private String email;
+    private String location;
```
[the remark this answers is not in the document]
Comment (rezigned) [reply src/main/java/co/omise/models/Customer.java:27]: Some object doesn't have `location` though. (e.g. `Bank Account` https://www.omise.co/bank-account-api)

File: src/main/java/co/omise/models/Document.java [hunk 5]
```diff
@@ -48,7 +48,7 @@ public void setLocation(String location) {
     public static class DeleteRequestBuilder extends RequestBuilder<Document> {
         private String disputeId;
         private String documentId;
-        public DeleteRequestBuilder(String documentId, String disputeId) {
+        public DeleteRequestBuilder(String disputeId, String documentId) {
```
[the remark this answers is not in the document]
Comment (rezigned) [reply src/main/java/co/omise/models/Document.java:51]: Noted krub 👌

File: src/main/java/co/omise/models/SystemInfo.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    public static class GetRequestBuilder extends RequestBuilder<SystemInfo> {
+
+        @Override
+        protected String method() {
+            return GET;
+        }
+
+        @Override
+        protected HttpUrl path() {
+            return buildUrl(Endpoint.API, "");
```
[the remark this answers is not in the document]
Comment (rezigned) [reply src/main/java/co/omise/models/SystemInfo.java:42]: It's new endpoint "/" :p

[the remark this answers is not in the document]
Comment (nuxzero) [reply src/main/java/co/omise/models/SystemInfo.java:42]: Really 😮

File: src/main/java/co/omise/models/ChargeStatus.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (danfowler) [reply src/main/java/co/omise/models/ChargeStatus.java:16]: Technically, it's a new status for `token.charge_status` and `source.charge_status`, but for `charge.status`, it hasn't changed.

[the remark this answers is not in the document]
Comment (nuxzero) [reply src/main/java/co/omise/models/ChargeStatus.java:16]: 👍

File: src/main/java/co/omise/models/Customer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (nuxzero) [reply src/main/java/co/omise/models/Customer.java:27]: 👍

**Pull request merged**